stdlib/public/core/OutputStream.swift

```diff
@@ -539,7 +539,12 @@ extension String: TextOutputStream {
   }
 
   public mutating func _writeASCII(_ buffer: UnsafeBufferPointer<UInt8>) {
-    self._guts.append(_StringGuts(buffer, isASCII: true))
+    if let smol = _SmallString(buffer) {
+      _guts.append(_StringGuts(smol))
+    } else {
+      _guts.reserveCapacity(_guts.utf8Count + buffer.count)
+      _guts.appendInPlace(buffer, isASCII: true)
+    }
```

Comment:
Were you finding this to produce better performance than what was there before?

Reply:
This API is only used in [FloatingPointTypes.swift.gyb](https://github.com/apple/swift/blob/c6cd4b07b2654cadc28cd16468e98bd2aac9e58b/stdlib/public/core/FloatingPointTypes.swift.gyb#L118-L127) for string interpolation.

My changes are made for safety rather than performance.

* https://github.com/apple/swift/pull/30780#issuecomment-611661725
  My initial naïve implementation (26dc7488d81cef11d59a5d3cdd50b711b3674a63) had some clear regressions:

  - StringComparison_abnormal
  - FloatingPointPrinting_Double_interpolated
  - FloatingPointPrinting_Float80_interpolated

* https://github.com/apple/swift/pull/30780#issuecomment-667249442
  My new implementation (0280b7abcaee934138f903bec2c2486d3722cb82) doesn't have those regressions.

Reply:
Out of curiosity, why did you choose `reserveCapacity` rather than `prepareForAppendInPlace`? `reserveCapacity` is not meant to handle append-like growth and though it might happen to put the string into a state for `appendInPlace` that's not part of its contract. All of this is handled by the existing append taking a StringGuts, hence why we were just calling it directly.

Since the contents of the append are non-escaping, should we just form an immortal guts?

Reply:
> Out of curiosity, why did you choose `reserveCapacity` rather than `prepareForAppendInPlace`?

* [`prepareForAppendInPlace`](https://github.com/apple/swift/blob/34650e8f3a1919cd337e25ec945346fb22ad8bec/stdlib/public/core/StringGutsRangeReplaceable.swift#L112-L152) is *private* API.

* [`repairUTF8`](https://github.com/apple/swift/blob/34650e8f3a1919cd337e25ec945346fb22ad8bec/stdlib/public/core/StringUTF8Validation.swift#L156-L204) is using `reserveCapacity` followed by `appendInPlace`.

Could the `prepareForAppendInPlace` call be moved, out of `append`, into `appendInPlace` and `_foreignAppendInPlace`?

> Since the contents of the append are non-escaping, should we just form an immortal guts?

My concern with putting a transient buffer into an immortal guts is safety. For example, when appending to an empty guts, would `self = immortalGuts` be allowed as a future optimization?

Reply:
@milseman I've tried the alternative (in commit 9d22cc0c2acdc30dc285e14cb054dda780440198), but I'm not sure how to interpret the benchmark results in https://github.com/apple/swift/pull/30780#issuecomment-670428613.